### DISCUSSION TITLE
voxel transformation ?

### DISCUSSION THREAD

**Discussion by StephaneKazmierczak** [pyvista/pyvista#1567] [Q&A] [answered]
Hello, 

I have been searching through your documentation and code but couldn't find, 
is it possible to do voxel transformation, such as rotation, flip or scale with pyvista ?

**Comment by tkoyama010** [comment, marked as answer]
You can use [rotate_x](https://docs.pyvista.org/core/dataset.html#pyvista.DataSet.rotate_x) , [rotate_y](https://docs.pyvista.org/core/dataset.html#pyvista.DataSet.rotate_y) , [rotate_z](https://docs.pyvista.org/core/dataset.html#pyvista.DataSet.rotate_z) methods to rotate meshs. There is not a method of flip and scale but you can easily do it using [points](https://docs.pyvista.org/core/dataset.html#pyvista.DataSet.points) property.
```python
import pyvista as pv

mesh = pv.Sphere()
# Scale
mesh.points *= 10.0
# Flip with x axis
mesh.points[:, 0] *= -1
# Flip with y axis
mesh.points[:, 1] *= -1
# Flip with z axis
mesh.points[:, 2] *= -1
```
But it is prefer to have a method for it. We will add it.

**Reply by tkoyama010** [reply]
Ref https://github.com/pyvista/pyvista/issues/1568

**Reply by StephaneKazmierczak** [reply]
Thanks you so much ! 
That is so cool, I missed that the voxels was a mesh object, I just though it was unstructuredgrid. I will check the code more, I am curious to see how you do the rotations (quaternion?)
I just got the functions you mentioned working on the voxel foot example (took me some time to render correctly, I was using win10/wsl/jupyter and had to switch to win10 only, but it could be a bad conda env...) 
I will try on my own voxel data next. 

Best,

**Reply by tkoyama010** [reply]
You can see how we rotate in the `axis_angle_rotation` function.
https://github.com/pyvista/pyvista/blob/1f7b911df803c7d088d46b6de3a65ebfd3726edc/pyvista/utilities/transformations.py#L5-L35

**Reply by StephaneKazmierczak** [reply]
Hello, I have spend a bit more time looking at the classes and documentation, but I don't find the way to create a mesh from my voxel data to apply function to them then return the update voxel data. Is is something supported by the library ? This the kind of voxel I am working with : 

```python
import numpy as np
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D

N1 = 10
N2 = 10
N3 = 10
ma = np.random.choice([0,1], size=(N1,N2,N3), p=[0.99, 0.01])

fig = plt.figure()
ax = fig.add_subplot(1,1,1, projection="3d")
ax.voxels(ma, edgecolor="k")

plt.show()
```

**Reply by RichardScottOZ** [reply]
Get the scalar array from your  voxels and voxels['data'] = sk_function(voxels['data'])
